Virtual point cloud + 3D map view + change visible point categories = crash

### What is the bug or the crash?

QGIS crashes when a virtual point cloud is viewed in the 3D map view and one or more point classifications is selected or deselected.  The exact number of (de)selects before the crash occurs varies; typically it's between one and half a dozen or so.

### Steps to reproduce the issue

This issue only appears when a virtual point cloud layer is being displayed in the 3D view.  To reproduce the issue:

1. Download and extract the project files: [point-cloud-crash-minimal.zip](https://github.com/user-attachments/files/19564733/point-cloud-crash-minimal.zip)
    This project contains two layers: a virtual point cloud and a "real" one.  Both are backed by the same physical file.
2. Download the point cloud to the same directory where you extracted the project files to.   (The point cloud file is too large [187.4 MB] to include in the project ZIP file.)  Either [download it directly](https://download-telecharger.services.geo.ca/pub/elevation/pointclouds_nuagespoints/SNCA/Eastern_Ontario_Part2_2021/1km185190498902021LEASTERNONTARIO.copc.laz) or run the following in the same directory:
    `wget https://download-telecharger.services.geo.ca/pub/elevation/pointclouds_nuagespoints/SNCA/Eastern_Ontario_Part2_2021/1km185190498902021LEASTERNONTARIO.copc.laz`
3. Open the project file `point-cloud-crash-minimal.qgs` in QGIS
4. In the Layers panel, deselect [or select] point categories ("Unclassified", "Ground", etc.) for the virtual point cloud (the first layer in the list).
5. After the visibility of one [or more, it varies] point categories is changed QGIS will crash.

The second layer in the project contains the same data as the first but as a "real" point cloud file.  Observe that categories in the second layer can be freely (de)selected in the non-virtual layer without issue.

---

The only apparent requirements to reproduce the issue are:

 - A visible virtual point cloud layer
 - The layer being displayed in a 3D view
 - Changing the active categories in the layer

The issue does not appear to be related to the size of the point cloud, the number of points lying within the 3D view, or the number of real files in the virtual point cloud.  It *may* also depend (at least to some degree) on the number of points being physically plotted to the 3D view [I observed the issue occurring sooner when the 3D view window is physically larger and it's my understanding that point culling is done prior to output; the issue is observed for all 3D view window sizes though].

### Versions

<table width='100%' align='center'><tr><td>QGIS version</td><td>3.43.0-Master</td></tr><tr><td>QGIS code revision</td><td><a href="https://github.com/qgis/QGIS/commit/be8ad62060">be8ad62060</a></td></tr><tr><td colspan="2">&nbsp;</td></tr><tr><td colspan="2"><b>Libraries</b></td></tr><tr><td>Qt version</td><td>5.15.3</td></tr><tr><td>Python version</td><td>3.10.12</td></tr><tr><td>GDAL version</td><td>3.4.1</td></tr><tr><td>PROJ version</td><td>8.2.1</td></tr><tr><td>EPSG Registry database version</td><td>v10.041 (2021-12-03)</td></tr><tr><td>GEOS version</td><td>3.10.2-CAPI-1.16.0</td></tr><tr><td>SQLite version</td><td>3.37.2</td></tr><tr><td>PDAL version</td><td>2.9.0</td></tr><tr><td>PostgreSQL client version</td><td>17.4</td></tr><tr><td>SpatiaLite version</td><td>5.0.1</td></td></tr><tr><td>QWT version</td><td>6.1.4</td></tr><tr><td>QScintilla2 version</td><td>2.11.6</td></tr><tr><td>OS version</td><td>Ubuntu 22.04.5 LTS</td></tr><tr><td colspan="2"><i>This copy of QGIS writes debugging output.</i></td></tr><tr><td colspan="2">&nbsp;</td></tr><tr><td colspan="2"><b>Active Python plugins</b></td></tr><tr><td>processing</td><td>2.12.99</td></tr><tr><td>db_manager</td><td>0.1.20</td></tr><tr><td>grassprovider</td><td>2.12.99</td></tr><tr><td>MetaSearch</td><td>0.3.6</td></tr></table>

The issue also occurs with the same version of QGIS built without debugging enabled.

### Supported QGIS version

- [x] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [x] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

The crash appears to be happening in a Qt signal handler so the stack trace isn't terribly useful in pinning down the source of the issue.  It does appear to be something related to the 3D viewer subsystem (either Qt's or QGIS').

Here's a stack trace of the crash:

```
QGIS died on signal 11
Extra Info File: /tmp/qgis-crash-info-2185417
[New LWP 2185429]
[New LWP 2185437]
[New LWP 2185438]
[New LWP 2185439]
[New LWP 2185440]
[New LWP 2185441]
[New LWP 2185442]
[New LWP 2185443]
[New LWP 2185444]
[New LWP 2185445]
[New LWP 2185446]
[New LWP 2185447]
[New LWP 2185448]
[New LWP 2185449]
[New LWP 2185456]
[New LWP 2185460]
[New LWP 2185465]
[New LWP 2185468]
[New LWP 2185544]
[New LWP 2185546]
[New LWP 2186013]
[New LWP 2186016]
[New LWP 2186017]
[New LWP 2186019]
[New LWP 2186020]
[New LWP 2186021]
[New LWP 2186022]
[New LWP 2186023]
[New LWP 2186025]
[New LWP 2186026]
[New LWP 2186027]
[New LWP 2186030]
[New LWP 2186033]
[New LWP 2186040]
[New LWP 2186125]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007fcd9b8ea42f in __GI___wait4 (pid=2186339, stat_loc=0x5589760e93e4, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30      ../sysdeps/unix/sysv/linux/wait4.c: No such file or directory.
[Current thread is 1 (Thread 0x7fcd873e3000 (LWP 2185417))]
#0  0x00007fcd9b8ea42f in __GI___wait4 (pid=2186339, stat_loc=0x5589760e93e4, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
        sc_ret = -512
        sc_cancel_oldtype = 0
#1  0x00005589716849fe in qgisCrash(int) ()
#2  0x00007fcd9b842520 in <signal handler called> () at /lib/x86_64-linux-gnu/libc.so.6
#3  __pthread_kill_implementation (no_tid=0, signo=11, threadid=140520714022912) at ./nptl/pthread_kill.c:44
        tid = <optimized out>
        ret = 0
        pd = 0x7fcd873e3000
        old_mask = {__val = {2672020178878332925, 140520130708848, 513, 59, 93459448058064, 0, 140516267666528, 140520714018008, 11, 140516265808082, 59, 0, 140516267666528, 140516264655903, 8026576034422810996, 7286952219023274853}}
        ret = <optimized out>
#4  __pthread_kill_internal (signo=11, threadid=140520714022912) at ./nptl/pthread_kill.c:78
#5  __GI___pthread_kill (threadid=140520714022912, signo=signo@entry=11) at ./nptl/pthread_kill.c:89
#6  0x00007fcd9b842476 in __GI_raise (sig=11) at ../sysdeps/posix/raise.c:26
        ret = <optimized out>
#7  0x00007fcd9b842520 in <signal handler called> () at /lib/x86_64-linux-gnu/libc.so.6
#8  0x00007fcd92bcdd1f in Qt3DCore::QNode::parentNode() const () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#9  0x00007fcd92bcf92b in Qt3DCore::QNodePrivate::_q_postConstructorInit() () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#10 0x00007fcd92bcfa2b in  () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#11 0x00007fcd92bc58be in Qt3DCore::QAspectManager::processFrame() () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#12 0x00007fcd92bc5c4a in  () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#13 0x00007fcda02f1793 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#14 0x00007fcda00b29dc in QAbstractAnimation::stop() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007fcda00b2bfc in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00007fcda00b1493 in QUnifiedTimer::updateAnimationTimers(long long) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#17 0x00007fcda00b3271 in QAnimationDriver::advanceAnimation(long long) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#18 0x00007fcda02e733f in QObject::event(QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#19 0x00007fcda0f6c713 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#20 0x00007fcd9d6d4137 in QgsApplication::notify(QObject*, QEvent*) () at /-/Projects/QGIS/qgis3/build/output/lib/libqgis_core.so.3.43.0
#21 0x00007fcda02b9e3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#22 0x00007fcda03123eb in QTimerInfoList::activateTimers() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#23 0x00007fcda0312d34 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#24 0x00007fcd92d20d3b in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#25 0x00007fcd92d762b8 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#26 0x00007fcd92d1e3e3 in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#27 0x00007fcda03130b8 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#28 0x00007fcda02b875b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#29 0x00007fcda02c0cf4 in QCoreApplication::exec() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#30 0x000055897168c87b in main ()
[Inferior 1 (process 2185417) detached]
gdb returned 0
Aborted
```

From a different crash:
```
#0  0x0000000000000000 in  ()
#1  0x00007ffff2cc2b5e in QMetaObject::cast(QObject const*) const () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#2  0x00007fffe515492b in Qt3DCore::QNodePrivate::_q_postConstructorInit() () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#3  0x00007fffe5154a2b in  () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#4  0x00007fffe514a8be in Qt3DCore::QAspectManager::processFrame() () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#5  0x00007fffe514ac4a in  () at /lib/x86_64-linux-gnu/libQt53DCore.so.5
#6  0x00007ffff2cf1793 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#7  0x00007ffff2ab29dc in QAbstractAnimation::stop() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#8  0x00007ffff2ab2bfc in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ffff2ab1493 in QUnifiedTimer::updateAnimationTimers(long long) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#10 0x00007ffff2ab3271 in QAnimationDriver::advanceAnimation(long long) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff2ce733f in QObject::event(QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#12 0x00007ffff396c713 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#13 0x00007fffefd50e0d in QgsApplication::notify(QObject*, QEvent*) (this=0x7fffffffd3b0, receiver=0x55555997e398, event=0x7fffffffcd30) at /-/Projects/QGIS/qgis3/QGIS/src/core/qgsapplication.cpp:628
#14 0x00007ffff2cb9e3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007ffff2d123eb in QTimerInfoList::activateTimers() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00007ffff2d12cec in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#17 0x00007fffe651bd3b in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#18 0x00007fffe65712b8 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#19 0x00007fffe65193e3 in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#20 0x00007ffff2d130b8 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#21 0x00007ffff2cb875b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#22 0x00007ffff2cc0cf4 in QCoreApplication::exec() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#23 0x000055555556f68a in main(int, char**) (argc=1, argv=0x7fffffffdda8) at /-/Projects/QGIS/qgis3/QGIS/src/app/main.cpp:1831
```

#### Additional errors seen at project load time

In the "General" log:
```
WARNING    Failed to calculate statistics of the point cloud Virtual point cloud
```

In the "Messages" log:
```
CRITICAL    Qt : QXcbConnection: XCB error: 3 (BadWindow), sequence: 1075, resource id: 21053764, major code: 40 (TranslateCoords), minor code: 0
CRITICAL    Qt : QXcbConnection: XCB error: 3 (BadWindow), sequence: 8978, resource id: 21053795, major code: 40 (TranslateCoords), minor code: 0
```

Possibly related, I noticed this while building a recent version https://github.com/qgis/QGIS/commit/29ae95fa84602f26ce6272054e7d75d52b994aa5 of the QGIS source:

```
[ 38%] Building CXX object src/core/CMakeFiles/qgis_core.dir/pointcloud/qgspointcloudattributemodel.cpp.o
/my/local/path/QGIS/src/core/mesh/qgsmeshvirtualdatasetgroup.cpp: In member function ‘bool QgsMeshVirtualDatasetGroup::calculateDataset() const’:
/my/local/path/QGIS/src/core/mesh/qgsmeshvirtualdatasetgroup.cpp:178:44: warning: writing 36 bytes into a region of size 24 [-Wstringop-overflow=]
  178 |     mDatasetMetaData[mCurrentDatasetIndex] = mCacheDataset->metadata();
In file included from /usr/include/x86_64-linux-gnu/qt5/QtCore/qbytearray.h:46,
                 from /usr/include/x86_64-linux-gnu/qt5/QtCore/qvariant.h:44,
                 from /usr/include/x86_64-linux-gnu/qt5/QtCore/qmetaobject.h:45,
                 from /usr/include/x86_64-linux-gnu/qt5/QtCore/QMetaEnum:1,
                 from /my/local/path/QGIS/src/core/qgis.h:22,
                 from /my/local/path/build-3.40-unmodified/src/core/CMakeFiles/qgis_core.dir/cmake_pch.hxx:5,
                 from <command-line>:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qarraydata.h:129:8: note: destination object ‘QTypedArrayData<QgsMeshDatasetMetadata>::<anonymous>’ of size 24
  129 | struct QTypedArrayData
      |        ^~~~~~~~~~~~~~~
```

--- a/src/3d/qgsvirtualpointcloudentity_p.cpp
+++ b/src/3d/qgsvirtualpointcloudentity_p.cpp
@@ -87,6 +87,15 @@ QgsVirtualPointCloudEntity::QgsVirtualPointCloudEntity(
   connect( provider(), &QgsVirtualPointCloudProvider::subIndexLoaded, this, &QgsVirtualPointCloudEntity::createChunkedEntityForSubIndex );
 }
 
+QgsVirtualPointCloudEntity::~QgsVirtualPointCloudEntity()
+{
+  qDeleteAll( mChunkedEntitiesMap );
+  mChunkedEntitiesMap.clear();
+
+  delete mOverviewEntity;
+  mOverviewEntity = nullptr;
+}
+
 QList<QgsChunkedEntity *> QgsVirtualPointCloudEntity::chunkedEntities() const
 {
   return mChunkedEntitiesMap.values();

--- a/src/3d/qgsvirtualpointcloudentity_p.h
+++ b/src/3d/qgsvirtualpointcloudentity_p.h
@@ -59,6 +59,9 @@ class QgsVirtualPointCloudEntity : public Qgs3DMapSceneEntity
     //! Constructs
     QgsVirtualPointCloudEntity( Qgs3DMapSettings *map, QgsPointCloudLayer *layer, const QgsCoordinateTransform &coordinateTransform, QgsPointCloud3DSymbol *symbol, float maxScreenError, bool showBoundingBoxes, double zValueScale, double zValueOffset, int pointBudget );
 
+    //! Destructs
+    ~QgsVirtualPointCloudEntity() override;
+
     //! This is called when the camera moves. It's responsible for loading new indexes and decides if subindex will be rendered as bbox or chunked entity.
     void handleSceneUpdate( const SceneContext &sceneContext ) override;
 

